File: legacy/python/src/mvmctl/api/inputs/_cp_input.py
"""CP input resolution — Input → Request → ResolvedCPInput."""

from __future__ import annotations

import logging
import os
from dataclasses import dataclass

from mvmctl.core._shared import Database
from mvmctl.core.config._service import SettingsService
from mvmctl.core.key._repository import KeyRepository
from mvmctl.core.key._resolver import KeyResolver
from mvmctl.core.key._service import KeyService
from mvmctl.core.ssh._cp import CPService
from mvmctl.core.vm._repository import VMRepository
from mvmctl.core.vm._resolver import VMResolver
from mvmctl.exceptions import CPError, SSHError
from mvmctl.models import VMInstanceItem

logger = logging.getLogger(__name__)


@dataclass
class CPInput:
    """Raw copy identifiers from CLI."""

    sources: list[str]
    dst: str
    user: str | None = None
    key: str | None = None
    force: bool = False


@dataclass
class ResolvedCPInfo:
    """Resolved connection info for one side of a copy operation."""

    identifier: str
    ip: str
    user: str
    key_path: str | None
    remote_path: str
    is_directory: bool | None = None
    total_bytes: int | None = None


@dataclass(frozen=True)
class ResolvedCPInput:
    """Fully resolved copy parameters."""

    direction: str  # "host_to_vm" | "vm_to_host" | "vm_to_vm"
    local_paths: list[str] | None = None
    src_info: ResolvedCPInfo | None = None
    dst_info: ResolvedCPInfo | None = None
    force: bool = False


class CPRequest:
    """Resolve CPInput against the database and filesystem."""

    def __init__(self, inputs: CPInput, db: Database) -> None:
        self._inputs = inputs
        self._db = db

    def resolve(self) -> ResolvedCPInput:
        """Resolve all inputs to explicit values."""
        sources = self._inputs.sources
        dst_vm, dst_path = CPService._parse_vm_path(self._inputs.dst)

        src_info: ResolvedCPInfo | None = None
        dst_info: ResolvedCPInfo | None = None
        local_paths: list[str] | None = None
        direction: str

        if len(sources) > 1:
            # Multi-source only works for host → VM
            if dst_vm is None:
                raise CPError(
                    "Multiple sources require a VM destination "
                    "(use vm_name:/path format)",
                    code="cp.multi_source_no_vm_destination",
                )
            direction = "host_to_vm"
            local_paths = sources
            if dst_vm is None:
                raise CPError(
                    "Internal error: destination VM not resolved",
                    code="cp.resolve_failed",
                )
            dst_info = self._resolve_vm_side(dst_vm, dst_path, is_source=False)
        else:
            # Single source — determine direction by parsing source
            src_path = sources[0]
            src_vm, src_remote_path = CPService._parse_vm_path(src_path)

            if src_vm and dst_vm:
                direction = "vm_to_vm"
            elif src_vm:
                direction = "vm_to_host"
            elif dst_vm:
                direction = "host_to_vm"
            else:
                raise CPError(
                    "At least one path must reference a VM "
                    "(use vm_name:/path format)",
                    code="cp.no_vm_specified",
                )

            if direction == "host_to_vm":
                # src is local, dst is VM
                local_paths = [src_path]
                if dst_vm is None:
                    raise CPError(
                        "Internal error: destination VM not resolved",
                        code="cp.resolve_failed",
                    )
                dst_info = self._resolve_vm_side(
                    dst_vm, dst_path, is_source=False
                )
            elif direction == "vm_to_host":
                # src is VM, dst is local
                if src_vm is None:
                    raise CPError(
                        "Internal error: source VM not resolved",
                        code="cp.resolve_failed",
                    )
                src_info = self._resolve_vm_side(
                    src_vm, src_remote_path, is_source=True
                )
                local_paths = [dst_path]
            elif direction == "vm_to_vm":
                # both are VMs
                if src_vm is None or dst_vm is None:
                    raise CPError(
                        "Internal error: source or destination VM not resolved",
                        code="cp.resolve_failed",
                    )
                src_info = self._resolve_vm_side(
                    src_vm, src_remote_path, is_source=True
                )
                dst_info = self._resolve_vm_side(
                    dst_vm, dst_path, is_source=False
                )

        return ResolvedCPInput(
            direction=direction,
            local_paths=local_paths,
            src_info=src_info,
            dst_info=dst_info,
            force=self._inputs.force,
        )

    def _resolve_vm_side(
        self, vm_ident: str, remote_path: str, is_source: bool
    ) -> ResolvedCPInfo:
        """Resolve a VM-side path to connection info."""
        vm = self._resolve_vm(vm_ident)
        user = self._resolve_user(vm)
        key_path = self._resolve_key(vm)

        if not vm.ipv4:
            raise CPError(
                f"VM '{vm_ident}' has no IP address assigned",
                code="cp.vm_no_ip",
            )

        return ResolvedCPInfo(
            identifier=vm_ident,
            ip=vm.ipv4,
            user=user,
            key_path=key_path,
            remote_path=remote_path,
        )

    def _resolve_vm(self, identifier: str) -> VMInstanceItem:
        """Resolve a VM by name, IP, MAC, or ID prefix."""
        repo = VMRepository(self._db)
        resolver = VMResolver(repo)
        try:
            return resolver.resolve(identifier)
        except Exception as e:
            raise CPError(
                f"Could not resolve VM '{identifier}': {e}",
                code="cp.vm_not_found",
            ) from e

    def _resolve_user(self, vm: VMInstanceItem) -> str:
        """Resolve SSH user."""
        if self._inputs.user is not None:
            return self._inputs.user
        if vm.ssh_user:
            return vm.ssh_user
        return str(SettingsService.resolve(self._db, "defaults.vm", "ssh_user"))

    def _resolve_key(self, vm: VMInstanceItem) -> str | None:
        """
        Resolve SSH private key path.

        Resolution order:
        1. If ``--key`` provided: try as registered key name, then as filesystem path.
        2. Check VM's stored ``ssh_keys``.
        3. Fall back to default keys.
        """
        key_repo = KeyRepository(self._db)
        key_resolver = KeyResolver(key_repo)

        if self._inputs.key is not None:
            key_str = self._inputs.key

            # Try as registered key name
            try:
                key_item = key_resolver.resolve(key_str)
                if key_item.private_key_path and os.path.exists(
                    key_item.private_key_path
                ):
                    return str(key_item.private_key_path)
            except Exception:
                pass

            # Try as filesystem path
            if os.path.exists(key_str) and os.path.isfile(key_str):
                key_service = KeyService(key_repo)
                with open(key_str) as f:
                    content = f.read()
                if key_service._is_private_key(content):
                    return key_str

            raise SSHError(
                f"Key '{key_str}' not found or is not a valid private key"
            )

        # Check VM's stored ssh_keys
        if vm.ssh_keys:
            for key_id in vm.ssh_keys:
                try:
                    key_item = key_resolver.by_id(key_id)
                    if key_item.private_key_path and os.path.exists(
                        key_item.private_key_path
                    ):
                        return str(key_item.private_key_path)
                except Exception:
                    continue

        # Fall back to default keys
        defaults = key_resolver.get_defaults()
        for key_item in defaults:
            if key_item.private_key_path and os.path.exists(
                key_item.private_key_path
            ):
                return str(key_item.private_key_path)

        return None

File: pkg/api/inputs/cp_input.go
package inputs

import (
	"context"
	"fmt"
	"mvmctl/internal/core/config"
	"mvmctl/internal/core/vm"
	"mvmctl/internal/core/vsock"
	"mvmctl/internal/infra"
	"mvmctl/internal/lib/model"
	"mvmctl/internal/lib/system"
	"mvmctl/pkg/errs"
)

// CPInput specifies c p input.
type CPInput struct {
	Sources []string `json:"sources" yaml:"src"`
	Dest    string   `json:"dest"    yaml:"dest"`
	Force   bool     `json:"force"   yaml:"force"`
}

// ResolvedCPInfo holds resolved target VM info for a copy operation.
type ResolvedCPInfo struct {
	Identifier  string
	RemotePath  string
	IsDirectory *bool
	TotalBytes  *int64
	Vsock       *model.VsockConfigItem
}

// ResolvedCPInput specifies resolved c p input.
type ResolvedCPInput struct {
	Direction  string
	LocalPaths []string
	SrcInfo    *ResolvedCPInfo
	DstInfo    *ResolvedCPInfo
	Force      bool
}

// Validate checks that the copy input has valid sources and destination.
func (i *CPInput) Validate() error {
	if len(i.Sources) == 0 {
		return errs.New(errs.CodeCPError, "At least one source path is required")
	}
	if i.Dest == "" {
		return errs.New(errs.CodeCPError, "A destination path is required")
	}
	return nil
}

// Resolve expands tilde paths and resolves VM identifiers to vsock config.
func (i *CPInput) Resolve(
	ctx context.Context,
	cfg *config.Service,
	vmRepo vm.Repository,
	vsockRepo vsock.Repository,
) (*ResolvedCPInput, error) {
	if err := i.Validate(); err != nil {
		return nil, err
	}
	sources := make([]string, len(i.Sources))
	for i, src := range i.Sources {
		sources[i] = system.ExpandTilde(src)
	}
	dstVM, dstPath := infra.ParseVMPath(i.Dest)
	srcVM, srcRemotePath := infra.ParseVMPath(sources[0])
	var (
		srcInfo, dstInfo *ResolvedCPInfo
		localPaths       []string
		direction        string
		err              error
	)
	// Determine direction.
	multiSource := len(sources) > 1
	switch {
	case multiSource && dstVM == "":
		return nil, errs.New(errs.CodeCPMultiSourceNoVMDest,
			"Multiple sources require a VM destination (use vm_name:/path format)")
	case multiSource || (srcVM == "" && dstVM != ""):
		direction = infra.DirectionHostToVM
		if multiSource {
			localPaths = sources
		} else {
			localPaths = []string{sources[0]}
		}
		dstInfo, err = resolveVMSide(ctx, dstVM, dstPath, vmRepo, vsockRepo)
		if err != nil {
			return nil, err
		}
	case srcVM != "" && dstVM == "":
		direction = infra.DirectionVMToHost
		localPaths = []string{dstPath}
		srcInfo, err = resolveVMSide(ctx, srcVM, srcRemotePath, vmRepo, vsockRepo)
		if err != nil {
			return nil, err
		}
	case srcVM != "" && dstVM != "":
		direction = infra.DirectionVMToVM
		srcInfo, err = resolveVMSide(ctx, srcVM, srcRemotePath, vmRepo, vsockRepo)
		if err != nil {
			return nil, err
		}
		dstInfo, err = resolveVMSide(ctx, dstVM, dstPath, vmRepo, vsockRepo)
		if err != nil {
			return nil, err
		}
	default:
		return nil, errs.New(errs.CodeCPNoVMSpecified,
			"At least one path must reference a VM (use vm_name:/path format)")
	}
	return &ResolvedCPInput{
		Direction:  direction,
		LocalPaths: localPaths,
		SrcInfo:    srcInfo,
		DstInfo:    dstInfo,
		Force:      i.Force,
	}, nil
}

// resolveVMSide resolves a VM-side path to vsock connection config.
func resolveVMSide(
	ctx context.Context,
	vmIdent, remotePath string,
	vmRepo vm.Repository,
	vsockRepo vsock.Repository,
) (*ResolvedCPInfo, error) {
	vmEntity, err := resolveVM(ctx, vmIdent, vmRepo)
	if err != nil {
		return nil, err
	}
	vsockCfg, err := vsockRepo.GetByVMID(ctx, vmEntity.ID)
	if err != nil {
		return nil, errs.Wrap(errs.CodeCPResolveFailed, err)
	}
	if vsockCfg == nil {
		return nil, errs.New(
			errs.CodeCPError,
			fmt.Sprintf("VM '%s' has no vsock configuration — ensure vsock device is enabled", vmIdent),
		)
	}
	return &ResolvedCPInfo{
		Identifier: vmIdent,
		RemotePath: remotePath,
		Vsock:      vsockCfg,
	}, nil
}

// resolveVM resolves a VM by name, IP, MAC, or ID prefix.
func resolveVM(ctx context.Context, identifier string, vmRepo vm.Repository) (*model.VMItem, error) {
	vmResolver := vm.NewResolver(vmRepo)
	vmEntity, err := vmResolver.Resolve(ctx, identifier)
	if err != nil {
		return nil, errs.NotFound(
			errs.CodeCPVMNotFound,
			fmt.Sprintf("Could not resolve VM '%s': %s", identifier, err.Error()),
		)
	}
	if vmEntity == nil {
		return nil, errs.NotFound(errs.CodeCPVMNotFound, fmt.Sprintf("VM '%s' not found", identifier))
	}
	return vmEntity, nil
}
